activitysim/abm/models/disaggregate_accessibility.py

```diff
@@ -0,0 +1,660 @@
+
+            uniform - Unweighted sample of N zones independent of each other.
+
+            uniform-taz - Unweighted sample of 1 zone per taz up to the N samples specified.
+
+            k-means - K-Means clustering is performed on the zone centroids (must be provided as maz_centroids.csv),
+                weighted by population. The clustering yields k XY coordinates weighted by zone population
+                for n-samples = k-clusters specified. Once k new cluster centroids are found, these are then
+                approximated into the nearest available zone centroid and used to calculate accessibilities on.
+
+                By default, the k-means method is run on 10 different initial cluster seeds (n_init) using using
+                "k-means++" seeding algorithm (https://en.wikipedia.org/wiki/K-means%2B%2B). The k-means method
+                runs for max_iter iterations (default=300).
+
+        """
+
+        # default_zone_col = 'TAZ' if not (self.network_los.zone_system == los.ONE_ZONE) else 'zone_id'
+        # zone_cols = self.model_settings["zone_id_names"].get("zone_group_cols", default_zone_col)
+        id_col = self.model_settings["zone_id_names"].get("index_col", "zone_id")
+        method = self.model_settings.get("ORIGIN_SAMPLE_METHOD")
+        n_samples = self.model_settings.get("ORIGIN_SAMPLE_SIZE", 0)
+
+        # Get weights, need to get households first to get persons merged.
+        # Note: This will cause empty zones to be excluded. Which is intended, but just know that.
+        zone_weights = self.land_use[
+            self.model_settings["ORIGIN_WEIGHTING_COLUMN"]
+        ].to_frame("weight")
+        zone_weights = zone_weights[zone_weights.weight != 0]
+
+        # If more samples than zones, just default to all zones
+        if n_samples == 0 or n_samples > len(zone_weights.index):
+            n_samples = len(zone_weights.index)
+            print("WARNING: ORIGIN_SAMPLE_SIZE >= n-zones. Using all zones.")
+            method = "full"  # If it's a full sample, no need to sample
+
+        if method and method == "full":
+            sample_idx = self.land_use.index
+        elif method and method.lower() == "uniform":
+            sample_idx = sorted(random.sample(sorted(self.land_use.index), n_samples))
+        elif method and method.lower() == "uniform-taz":
+            # Randomly select one MAZ per TAZ by randomizing the index and then select the first MAZ in each TAZ
+            # Then truncate the sampled indices by N samples and sort it
+            sample_idx = (
+                self.land_use.sample(frac=1)
+                .reset_index()
+                .groupby("TAZ")[id_col]
+                .first()
+            )
+            sample_idx = sorted(sample_idx)
+        elif method and method.lower() == "kmeans":
+            # Performs a simple k-means clustering using centroid XY coordinates
+            centroids_df = pipeline.get_table("maz_centroids")
```

Comment:
Seems `kmeans` will not work with one-zone system. Can we add `assert` before this?

Reply:
Added assertion for 2-zone systems. 1-zone could be implemented in future but may not be needed if aggressive sampling isn't typically requires for one-zone systems.